### DISCUSSION TITLE
How to started the server on a virtual python environment?

### DISCUSSION THREAD

**Discussion by feiwuu638** [fastapi/fastapi#7510] [Questions] [answered]
Hi.

I imported a package named `pgmpy` in the api, but the package is in my virtual environment. So when I ran `uvicorn main:app` in the command lines, it raised the `ModuleNotFoundError: No module named 'pgmpy'`. **I wondered how I can set the python environment of the api.**

Thank you.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI
# the pgmpy module is not used in this example, but it is needed in my real api
from pgmpy.models import BayesianModel

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}
```

### Description

* Run `uvicorn main:app` in the command lines.
* It raised the `ModuleNotFoundError: No module named 'pgmpy'`.
* I expected to started the server successfully based on the virtual python environment.

### Environment

* OS: [Windows]:
* FastAPI Version [e.g. 0.58.0]:
* Python version: 3.7.3

**Comment by zihaooo** [comment]
You should do all the stuff in the same environment. 
``` bash
$ source activate my-env
(my-env) $ pip install -r requirements.txt
(my-env) $ uvicorn main:app --reload
```

**Comment by gil-obradors** [comment]
Hi,
This question is about python basics : package imports

Your output says that module can t be found, so maybe not instaled in
virtual enviorment, fastagi not executed into virtualenv scope /virtual
enviorment not activated

I think the second, but you can confirm  with
* activate virtenv
* python
    - import pgmpy


If it fails, module not instaled in virtualenv

You can confirm packages installed with `pip list`

Good luck

El dg., 28 de juny 2020, 9:57, feiwuu638 <notifications@github.com> va
escriure:

**Comment by feiwuu638** [comment]
@zihaooo I tried to activate the virtual environment in the command lines, and then run the `uvicorn main:app`, but it still doesn't work.

**Comment by feiwuu638** [comment]
@gil-obradors The module is installed in the virtual environment, I can confirm that. And I run the fastapi in the activated virtual environment but it failed in the same way.

**Comment by zihaooo** [comment]
I checked again the pgmpy official document https://github.com/pgmpy/pgmpy#installation  If you followed these steps, you should have a conda (Anaconda/Miniconda), so you can create/use the environment by following commands:
```bash
$ conda create -n my-env python==${THE_VERSION_YOU_NEED}
$ conda activate my-env
(my-env) $ pip install -r requirements.txt  # you should put uvicorn in your requirements.txt
(my-env) $ uvicorn main:app --reload
```
You can also check where is the uvicorn binary:
```bash
(my-env) $ which uvicorn
```

You can check all the package installed in your env:
```bash
(my-env) $ pip list
```

**Remember**: 

1. Execute these commands in your environment.
2. uvicorn should be installed in your environment.
3. Don't use conda install and pip install at same time (in the same environment).

**Comment by feiwuu638** [comment, marked as answer]
@zihaooo I think the problem is solved. Originally, I have uvicorn installed in both env1 and env2, and pgmpy installed only in env2. In that case, I activated the env2 and ran `uvicorn main:app` but it didn't work. Receiving your advice, I uninstall the uvicorn from env1 and only keep it in env2. Then I executed the same comands in env2 and it works. Thank you very much for your kind help.

**Comment by zihaooo** [comment]
@feiwuu638 Glad to help you solve this problem. Then, you can close this issue.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:

**Comment by camilaagw** [comment]
@feiwuu638  thank you for sharing you solution! I had the same issue. I was working with only one venv and I fixed it by uninstalling `uvicorn` from the system Python interpreter

**Comment by greatvovan** [comment]
> This question is about python basics : package imports

I would slightly disagree with that. The problem is that when you install uvicorn, it creates eponymous command, which gets associated with the particular environment (or global interpreter). I am still not sure how to instruct the command to jump between envs, rather than reinstalling the command.
